Allow to Disable Moving Feature in NxTransferListHalf - RSC-1254

cindyko226 opened the pull request:
https://issues.sonatype.org/browse/RSC-1254

EmilyBermudez88 commented:
I think it's a good idea to add an example of the new disabledItemMove prop in the gallery

EmilyBermudez88 commented on gallery/src/components/NxTransferListHalf/NxTransferListHalfPage.tsx:
```diff
@@ -108,6 +108,15 @@ const NxTransferListPage = () =>
+              <NxTable.Cell>No</NxTable.Cell>
+              <NxTable.Cell></NxTable.Cell>
+              <NxTable.Cell>
+                Set to true to disable moving item feature and hide the move icon on each item.
```
Maybe something like : 'When true it disable’s the ability to transfer or remove items by hiding the item’s icon.'

rpokorny commented:
Rather than introducing a new prop, would it make sense to just make `onItemChange` optional, and to not show the icons when it is not set?

cindyko226 commented:
> I think it's a good idea to add an example of the new disabledItemMove prop in the gallery

Sure. Added here 
914f60cb038492809ddb496129f90f9b0036bca8
aaf191ce1d866ecba96aa6ffa3c95f93c7bb490e

cindyko226 commented:
> Rather than introducing a new prop, would it make sense to just make `onItemChange` optional, and to not show the icons when it is not set?

Yes, it makes sense to me since the `onItemChange` prop is not necessary when disabling the transfer feature. 
Updated here aaf191ce1d866ecba96aa6ffa3c95f93c7bb490e, thanks.

rpokorny commented on gallery/src/components/NxTransferListHalf/NxTransferListHalfDisableTransferExample.tsx:
```diff
@@ -0,0 +1,31 @@
+                             filterValue={filter}
+                             onFilterChange={setFilter}
+                             showMoveAll={false}
+                             onMoveAll={() => {}}
```
It seems a little goofy that `showMoveAll` and `onMoveAll` need to be provided in this case.  For that matter, it seems a little goofy that `onMoveAll` needs to be provided any time `showMoveAll` is false.  IMO those two props should also be made optional and excluded from this example.

cindyko226 replied:
Updated here 8a6881fef17ec3292383fa277f21d76019b90b93.
Further on this, since there won't be selected/unselected items in this case, should `isSelected` prop be optional as well?

rpokorny replied:
> Further on this, since there won't be selected/unselected items in this case, should isSelected prop be optional as well?

Sure

cindyko226 replied:
Updated here 8b6823bcb2af5335fecb102363cf56f47854aed8

rpokorny commented:
The items still have hover and click styles when moving is disabled, even though clicking them doesn't do anything.  Should those styles be disabled?

rpokorny commented:
Needs a visual test.  Also I think the original use case was to have reordering but not removal.  So we should have an example (and visual test) for that too.

cindyko226 commented:
> The items still have hover and click styles when moving is disabled, even though clicking them doesn't do anything. Should those styles be disabled?

Checked with design @JasonDSwearingen, in this case the item should not be clickable or have interaction states while tooltip still shows on hover when display name is truncated.
Updated sketch mockups: https://www.sketch.com/s/2393264e-1233-4909-99c8-a7dda02588cd/a/m1095Gz

Those styles should be disabled, updated here 9370018954ef06e3f5b78538ed371cab13bea0e5

rpokorny commented on gallery/src/components/NxTransferListHalf/NxTransferListHalfDisableTransferExample.tsx:
```diff
@@ -0,0 +1,28 @@
+import { map, range, prepend } from 'ramda';
+import { faArrowsAltH } from '@fortawesome/free-solid-svg-icons';
+
+const Items: DataItem<number>[] = prepend(
```
There's no reason this should start uppercase

```suggestion
const items: DataItem<number>[] = prepend(
```

rpokorny commented on gallery/src/components/NxTransferListHalf/NxTransferListHalfPage.tsx:
```diff
@@ -69,16 +71,19 @@ const NxTransferListPage = () =>
               <NxTable.Cell>boolean</NxTable.Cell>
-              <NxTable.Cell>Yes</NxTable.Cell>
+              <NxTable.Cell>No</NxTable.Cell>
               <NxTable.Cell></NxTable.Cell>
```
Optional props need to have their default value specified when applicable

cindyko226 replied:
Updated here 9cf90c48d024ff1921bec3ca8088ae74e46aa821

rpokorny commented on lib/src/components/NxTransferListHalf/NxTransferListHalf.tsx:
```diff
@@ -40,14 +40,17 @@ function _TransferListItem<T extends string | number = string>(props: TransferLi
         'nx-transfer-list__item--with-reordering': !!showReorderingButtons
       },
+      {
+        'nx-transfer-list__moving--disabled': !onChangeProp
```
This isn't correct BEM.  I'd call it `nx-transfer-list__item--moving-disabled`.  Or put a `.nx-transfer-list__half--moving-disabled` class on the `.nx-transfer-list__half`

rpokorny replied:
> This isn't correct BEM. I'd call it nx-transfer-list__item--moving-disabled. Or put a .nx-transfer-list__half--moving-disabled class on the .nx-transfer-list__half

Better yet make a class for the opposite case: a `.nx-transfer-list__item--movable` or something.  Writing the CSS for that should be easier

cindyko226 replied:
Ah I miss the item.
Thanks for the suggestions, updated here 6f2d800213933eb20a58d91c8b7f579d8ef176b8.

rpokorny commented on lib/src/components/NxTransferListHalf/__tests__/NxTransferListHalf.test.tsx:
```diff
@@ -171,9 +168,23 @@ describe('NxTransferListHalf', function() {
-        withoutSelected = getMounted({ items: [{ id: 1, displayName: 'foo' }] });
+    const onItemChange = jest.fn(),
+        withSelected = getMounted({ items: [{ id: 1, displayName: 'foo' }], isSelected: true, onItemChange }),
+        withoutSelected = getMounted({ items: [{ id: 1, displayName: 'foo' }], onItemChange });
```
Should test `isSelected = false` as well

cindyko226 replied:
Updated here 4dc735f5e5008c35a9245561f423c697ce343183.

cindyko226 commented:
> Needs a visual test. 

Added here 
d658e734d93d6adc1165298a9550120c4eaa62ed
a73bfb86162ef154bfc35f817c42ee5a6e804235.

> Also I think the original use case was to have reordering but not removal. So we should have an example (and visual test) for that too.

Added here 
8802b01b53ca01d321a4347d69d1d2755b8003fc
This is the combination of disable transfer and ordering example, not sure if it is necessary to also have the visual tests for interaction states on each item and move up/down buttons, let me know your opinions, thanks.

rpokorny commented on lib/src/components/NxTransferListHalf/NxTransferListHalf.tsx:
```diff
@@ -148,7 +148,7 @@ export default function NxTransferListHalf<T extends string | number = string>(p
           </button>
         }
-        <div className="nx-transfer-list__item-list">
+        <div className="nx-transfer-list__item-list" tabIndex={onItemChange || allowReordering ? undefined : 0}>
```
> Add the tabIndex here to meet the a11y requirement that scrollable region must have keyboard access

Great.  I think that's worth adding a code comment

cindyko226 replied:
Updated here. 0211efb136eb72fcc99e81fdd4fd32ad193237cd

rpokorny commented on gallery/src/components/NxTransferListHalf/NxTransferListHalfPage.tsx:
```diff
@@ -90,8 +99,8 @@ const NxTransferListPage = () =>
-              <NxTable.Cell>Yes</NxTable.Cell>
-              <NxTable.Cell></NxTable.Cell>
+              <NxTable.Cell>No</NxTable.Cell>
+              <NxTable.Cell>false</NxTable.Cell>
```
IMO we should actually make this default to true.  In cases where you have just a single `NxTransferListHalf`, it would presumably be holding things that are selected.

rpokorny commented on lib/src/components/NxTransferListHalf/NxTransferListHalf.scss:
```diff
@@ -104,22 +104,29 @@ $-visible-item-count: 10;
-  cursor: pointer;
   flex: auto;
   padding: var(--nx-spacing-2x);
+  padding-left: var(--nx-spacing-4x);
```
Why is this 4x? Shouldn't it be 2x whether the icon is there or not?

rpokorny replied:
> Why is this 4x? Shouldn't it be 2x whether the icon is there or not?

I see that this does match the mockup.  However I recall a discussion with the designers where I believe they confirmed the spacing would be the same with and without the icon.  So unless you've doubled checked this with them already, I'm wondering if it's a mistake in the mockup.

cindyko226 replied:
I confirmed with designs @JasonDSwearingen, the padding would be 4x when rendered without the icon.

rpokorny commented on gallery/visualtests/nxTransferListHalf.js:
```diff
@@ -111,5 +116,21 @@ describe('NxTransferListHalf', function() {
+    it('looks right', simpleTest(disableTransferListSelector));
+    it('does not put a dark border on hovered items',
+        hoverTest(disableTransferListSelector, secondDisableTransferItemSelector));
+    it('does not put a blue border and glow on focused items',
+        focusTest(disableTransferListSelector, secondDisableTransferItemSelector));
+    it('does not put a blue border and blue glow on focused+hovered items',
+        focusAndHoverTest(disableTransferListSelector, secondDisableTransferItemSelector));
```
Given that the user can't actually focus the items I don't think these two tests are necessary

rpokorny commented on lib/src/components/NxTransferListHalf/NxTransferListHalf.tsx:
```diff
@@ -40,14 +40,17 @@ function _TransferListItem<T extends string | number = string>(props: TransferLi
+    onChangeProp?.(evt.currentTarget.checked, id);
   }
 
   const classes = classnames(
       'nx-transfer-list__item',
       {
         'nx-transfer-list__item--with-reordering': !!showReorderingButtons
       },
+      {
+        'nx-transfer-list__item--movable': !!onChangeProp
+      }
   );
```
```suggestion
  const classes = classnames('nx-transfer-list__item', {
    'nx-transfer-list__item--with-reordering': !!showReorderingButtons
    'nx-transfer-list__item--movable': !!onChangeProp
  });
```

rpokorny commented on lib/src/components/NxTransferListHalf/NxTransferListHalf.tsx:
```diff
@@ -106,7 +108,7 @@ export default function NxTransferListHalf<T extends string | number = string>(p
         showMoveAll,
         onMoveAll,
-        isSelected,
+        isSelected = true,
```
Because we also allow null on optional props, this syntax is unfortunately not usually useful for us - it only catches `undefined`.  You should add unit tests for when `isSelected` is `null`, and you'll see that they don't pass.

cindyko226 replied:
Thanks for catching it, updated here 15fa580e13a126a1104f8bf8e856d24e4e1ed3df.
